WIP: PBS-to-Slurm adaptation (#306)
hpcleuven/VscDocumentation: source/leuven/slurm-accounting.rst

Slurm Accounting
========================

> The :ref:`wICE <wice_t2_leuven>` cluster uses the Slurm Account Manager (SAM) 

Author: of course it does :-)
fixed via `d24f13d`.

for credit accounting.
The reason is that having a valid Slurm account with sufficient credits in it
is mandatory for running jobs on KU Leuven Tier-2 machines. 
We developed so called SAM-commands to mimic the behavior of previously used 
MAM-commands (from the MOAB account manager). 
From the user point of view it should mostly be enough to change the prefix ``mam`` 

> in the accounting command into ``sam``.

Author: thanks for providing a text; I used yours.

The accounting system is very similar to a regular bank. 
Individual users have accounts that will be charged for the jobs they run. 
However, the number of credits on such accounts is fairly small, so research projects 

> will typically have one or more project accounts associated with them. 

Reviewer: I am not entirely sure what you mean here. Perhaps:

The accounting system is somewhat similar to a regular bank. A research group typically has one or more credit accounts, for instance a separate credit account for each project. When credits are purchased, they are depositied in a credit account. All users that are a member of such a credit account can withdraw credits from it by running jobs.
A users can also request introduction credits, in that case a credit account will be opened specifically for this user with a limited amount of free credits.

Author: Thanks for providing an improved text.
fixed via `d24f13d`.

Users that are project members can have their project-related jobs charged to such a 
project account. 

> In this how-to, the technical aspects of accounting are explained.

Author: I partly used your proposed text, because the same paragraph had changed as proposed by Steven.

Checking an account balance
---------------------------

> Since no calculations can be done without credits, it is quite useful to determine the 

Reviewer 2: This is not true for certain partitions (e.g. the interactive partition on wICE, as well as the dedicated partitions on wICE and Genius).

Author: true. I changed to text to this:
"It is a good practice to check the balance of your credit project(s) from time to time."

amount of credits at your disposal. 
This can be done by executing the following command on any KU Leuven Tier-2 node::

   $ sam-balance

> This will provide an overview of the balance on the user's personal account, as well as 
> on all project accounts the user has access to.

Reviewer 2: ```suggestion
This will show the balance of each credit account the user has access to.
```

Running jobs: accounting workflow
---------------------------------

When a job is submitted using ``sbatch`` or ``srun``, it has to be charged against a 
project account.
The name of the project has to be specified as an option. 
If a user requests introductory credits, the assigned project name carries the VSC-id

> of the user, e.g. ``intro_vsc3....``, where the dots ``.`` signify the user's uniqe VSC-id.

Reviewer: I think it would be clearer to write:

of the user, e.g. ``intro_vsc39999``. In the following commands, you will need to replace `vsc39999` with your own VSC-id.

Author: fixed via `d24f13d`.

> When a job is submitted using ``sbatch`` or ``srun``, it has to be charged against a 
> project account.
> The name of the project has to be specified as an option. 
> If a user requests introductory credits, the assigned project name carries the VSC-id
> of the user, e.g. ``intro_vsc3....``, where the dots ``.`` signify the user's uniqe VSC-id.

Reviewer 2: In general, I think it is best to avoid the term "project" / "project account", because it is unnecessary and carries the risk of additional confusion.

```suggestion
For every job that gets submitted, the name of a credit account needs to be specified.
If you for example have been granted introduction credits, the corresponding credit
account will be named ``intro_vscxxxxx`` (with ``vscxxxxx`` referring to your VSC username).

Submitting a batch job can then look as follows::
```

   $ sbatch -A lp_my_project run-job.slurm
   or

>    $ sbatch -A intro_39999 run-job.slurm

Reviewer: The formatting is off here, I think you forgot the `::` at the end of the preceding sentence.

Author: fixed via `d24f13d`.

>    $ sbatch -A intro_39999 run-job.slurm

Reviewer 2: ```suggestion
   $ sbatch -A intro_vsc39999 run-job.slurm
```

If the account to be charged, i.e., ``lp_my_project``, has insufficient credits for the 

> job, the user receives a warning at this point, and the job will never enter the queue.

Reviewer 2: ```suggestion
job, the user receives a warning at this point, and the job will not enter the queue.
```
("never" suggests that there could be multiple future times at which the job may still somehow be eligible for entering the queue, whereas in reality it is just a "one-off" thing: either the job enters the queue when it is submitted, or it gets rejected)

Author: Alright; it reads like this now:
"If the account to be charged, i.e., ``lp_my_project``, has insufficient credits for the 
job, the user receives a warning at this point, and the job will start until the account
is topped up with sufficient credits."

Obtaining an overview of transactions
-------------------------------------

A bank provides an overview of the financial transactions on your accounts under the 
form of statements. 
Similarly, the job accounting system provides statements that give the user an overview 
of the cost of each individual job. 
The following command will provide an overview of all transactions on all accounts the 
user has access to, as well as a summary of the credit usage at the top::

     $ sam-statement

It is more convenient to filter this information so that only specific projects are 
displayed and/or information for a specific period of time, e.g.::

   $ sam-statement -A lp_my_project -s 2023-01-01 -e 2023-01-31

This will show the transactions on the account for the ``lp_my_project`` project for 
the month January 2023.

> If you are only interested in the overview of user's transactions, and don't require 
> the actual balance information, ``sam-list-usagerecords`` provides a much faster 
> alternative for a summarized statement::

Reviewer 2: ```suggestion
If you are only interested in the overview of your past credit usage, and don't require 
the actual balance information, ``sam-list-usagerecords`` provides a much faster 
alternative for a summarized statement::
```

   $ sam-list-usagerecords -A lp_my_project -s 2023-01-01 -e 2023-01-31

.. note::

   - It takes quite a while to compute such statements, so **please be patient**

>    - The full statements are only visible to the project leaders. 
>      Individual users can only see their usage, but not the whole project usage

Reviewer 2: ```suggestion
   - Individual users can only see their own usage and not that of other users of the same credit account.
     The latter is only available to users who have been given a Coordinator role.
```

Author: using your text.

>    - All ``sam``-commands provide help with ``-h`` option

Reviewer: with -> by running them with the

Author: fixed via `d24f13d`.

> Slurm Accounting

Author: sure.

All users that are a member of such a credit account can withdraw credits from it 
by running jobs.
A users can also request introduction credits, in that case a credit account will 
be opened specifically for this user with a limited amount of free credits.
Hence, it is possible for users to have access to multiple credit accounts.

In this page, the technical aspects of accounting and some relevant commands are explained.

Checking an account balance
---------------------------

It is a good practice to check the balance of your credit project(s) from time to time.
This can be done by executing the following command on any KU Leuven Tier-2 node::

   $ sam-balance

This will show the balance of each credit account the user has access to.

Running jobs: accounting workflow
---------------------------------

For every job that gets submitted, the name of a credit account needs to be specified.
If you for example have been granted introduction credits, the corresponding credit
account will be named ``intro_vscxxxxx`` (with ``vscxxxxx`` referring to your VSC username).
Submitting a batch job can then look as follows::

   $ sbatch -A lp_my_project run-job.slurm
   or

>    $ sbatch -A intro_3xxxx run-job.slurm

Reviewer 2: ```suggestion
   $ sbatch -A intro_vsc3xxxx run-job.slurm
```

If the account to be charged, i.e., ``lp_my_project``, has insufficient credits for the 

> job, the user receives a warning at this point, and the job will start until the account

Reviewer 2: ```suggestion
job, the user receives a warning at this point, and the job will not start until the account
```

is topped up with sufficient credits.

Obtaining an overview of transactions
-------------------------------------

A bank provides an overview of the financial transactions on your accounts under the 
form of statements. 
Similarly, the job accounting system provides statements that give the user an overview 
of the cost of each individual job. 
The following command will provide an overview of all transactions on all accounts the 
user has access to, as well as a summary of the credit usage at the top::

>      $ sam-statement

Author: Now it reads like this:

The following command will provide an overview of all transactions on all accounts
that the user has access to, as well as a summary of the credit usage at the top::

`     $ sam-statement -u <vsc-account>`

>      $ sam-statement

Author: you're right. I have fixed that.
